File: 2017/21.py

import math, copy, re, hashlib
import itertools as it
import numpy as np

def rl(arr):
	return range(len(arr))

def get_bin(s):
	sv = 0
	l = len(s)
	for i in range(l-1, -1, -1):
		sv |= (1 << (i)) if s[i] == '#' else 0
	return sv

def get_mat(n, sz):
	s = ''
	while n:
		if n & 1: 
			s += '#'
		else:
			s += '.'
		n >>= 1
	s += (sz*sz - len(s)) * '.'
	ret = []
	for i in range(0, len(s), sz):
		ret.append(s[i:i+sz])
	return np.matrix([list(i) for i in ret])



def parse(rule):
	[l, r] = list(map(lambda x: x.strip(), rule.split('=>')))
	lv = rv = 0
	l = l.replace('/', '')
	r = r.replace('/', '')
	return get_bin(l), get_bin(r)

TWOS = 6

def bin_search(rules, s):
	lo = 0
	hi = len(rules)
	while lo < hi:
		md = lo + (hi - lo) // 2
		if rules[md][0] == s:
			return md
		elif rules[md][0] < s:
			lo = md + 1
		else:
			hi = md
	return None

def find_rule(sub, rules):
	l = len(sub)
	rule_list = rules[:TWOS] if l == 2 else rules[TWOS:]
	rule = None

	for _ in range(4):
		s = ''.join([''.join(row) for row in np.array(sub)])
		s = get_bin(s)
		rule = bin_search(rule_list, s)
		if rule is None:
			s = ''.join([''.join(row[0]) for row in np.array([sub[i] for i in range(len(sub)-1, -1, -1)])])
			s = get_bin(s)
			rule = bin_search(rule_list, s)
		if rule is not None:
			break
		sub = np.rot90(sub)
	return rule_list[rule]


def enhance(pattern, rules):
	l = len(pattern)
	step = 2 if l % 2 == 0 else 3
	sz = (l//step) * (step + 1)
	ret = np.array(range(sz*sz), dtype=str).reshape(sz, sz)
	for i in range(0, l//step):
		for j in range(0, l//step):
			sub = pattern[i*step:i*step+step,j*step:j*step+step]
			r = find_rule(np.matrix(sub), rules)
			mat = get_mat(r[1], step + 1)
			ret[i*(step+1):i*(step+1)+step+1,j*(step+1):j*(step+1)+step+1] = mat
	return ret

def part_1(data):

	rules = [parse(r) for r in data]
	
	pattern = ['.#.','..#','###']
	pattern = np.matrix([list(i) for i in pattern])

	for _ in range(18):
		print(_)
		pattern = enhance(pattern, rules)
		# print(pattern)
	cnt = 0
	for r in pattern:
		for c in r:
			if c == '#':
				cnt += 1
	print(cnt)

	print('END OF PART1')
	return

def part_2(data):
	print('END OF PART2')
	return 


if __name__ == '__main__':
	with open('21_input') as f:
		data = f.read()
		data = data.split('\n')
		# data = list(map(int, data.split()))


	part_1(copy.deepcopy(data))
	part_2(copy.deepcopy(data))
	

File: 2017/21.js

const fs = require('fs');
const { exec } = require("child_process");
const { ALPHA_L, ALPHA_U, D4, D6, D8, MOD } = require("../lib");
const { joseph, findCycle } = require("../lib");
const { cache } = require("../lib");
const { bin, float, hex, int, num, oct } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { copy,	entries, in_, inBB, keys, sort, values } = require("../lib");
const { Counter, DD, empty, FastQueue, PriorityQueue, Queue, set, Stack } = require("../lib");
const { areaInt, circumference, manDist, shoelace } = require("../lib");
const { ord, chr, count, debug, disp, disp3, crt, gcd, lcm, modPow, modPowBig, modInv, mod, prod, prodBig, randint, sum, sumBig, transpose } = require("../lib");
const { digits, ints,	floats,	singles,	words,	lines,	table,	groups,	getGroups,	groupsWith,	parse,	parseLine, scanf } = require ('../lib');
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { isSuperset, or, and, xor, sub } = set;
const { getExecStr } = require("../lib/post");
const { combinations, combinations_with_replacement, next_permutation, product } = require("../lib");

const getKey = (G) => {
	return G.map(r => r.join("")).join("/")
}
const fromKey = (s) => {
	return s.split("/").map(r => r.split(""));
}

const unit = (G) => G;
const rotate = (G) => {
	const nG = new Array(G.length).fill(0).map(r => new Array(G.length).fill("."));
	for (let i = 0; i < G.length; ++i) {
		for (let j = 0; j < G.length; ++j) {
			nG[j][G.length - i - 1] = G[i][j];
		}
	}
	return nG;
};
const flip = (G) => {
	for (let i = 0; i < G.length; ++i) {
		G[i].reverse();
	}
	return G;
};

const sides = [unit, rotate, rotate, rotate, (G) => flip(rotate(G)), rotate, rotate, rotate];

const split = (P, MAP, n) => {
	const step = P.length / n;
	const sz = step * (n + 1);
	const res = new Array(sz).fill(0).map(r => new Array(sz).fill("."));
	drange(P.length / n)((i, j) => {
		let G = P.slice(n * i, n * (i + 1)).map(r => r.slice(n * j, n * (j + 1)));
		iter(sides, f => {
			G = f(G);
			if (getKey(G) in MAP) return true;
		})
		const nG = fromKey(MAP[ getKey(G) ]);
		drange(nG.length)((di, dj) => {
			res[(n+1) * i + di][(n+1) * j + dj] = nG[di][dj];
		})
	})
	return res;
}

function part1(data) {

	let res = 0;
	data = scanf(data, "%$ => %$")
	const MAP = {};
	iter(data, ([from, to]) => {
		MAP[from] = to;
	})
	let P = [".#.", "..#", "###"].map(s => s.split(""))
	range(5)(_ => {
		if (P.length % 2 === 0) {
			P = split(P, MAP, 2);
		} else {
			P = split(P, MAP, 3);
		}
	})
	iter(P, r => {
		iter(r, c => {
			res += c === '#';
		})
	})

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART1");
	return;
}

function part2(data) {

	let res = 0;
	data = scanf(data, "%$ => %$")
	const MAP = {};
	iter(data, ([from, to]) => {
		MAP[from] = to;
	})
	let P = [".#.", "..#", "###"].map(s => s.split(""))
	range(18)(_ => {
		if (P.length % 2 === 0) {
			P = split(P, MAP, 2);
		} else {
			P = split(P, MAP, 3);
		}
	})
	iter(P, r => {
		iter(r, c => {
			res += c === '#';
		})
	})

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART2");
	return;
}

function main() {
	let data = fs.readFileSync("21_input").toString("utf-8");

	part1(data);
	part2(data);
	process.exit(0);
}

main();
